### ISSUE TITLE
[(Brainde)Code Sprint] Normalise models implementations

### ISSUE THREAD

**Issue by PierreGtch** [braindecode/braindecode#457]
**So far**, all the Braindecode models have been implemented independently, either inheriting from `torch.nn.Module` or from `torch.nn.Sequential`. Yet, they all receive inputs of the same shape (batch, channels, time), and they share many parameters and processing steps:
- *Shared by all models*:
    - number of classes (parameter)
    - number of channels (parameter)
    - fully-connected + softmax at the end (processing step)
- *Shared by some models*:
    - temporal length of the input (parameter)
    - sampling frequency (parameter)
    - dropout probability (parameter)
    - temporal and spatial filter at the beginning (processing step)

---

**The goal** of this meta-issue is to exploit these similarities to build a more unified framework. Specifically, this means:
- Create base [abstract](https://docs.python.org/3/library/abc.html) models from which all the Braindecode models should inherit;
- Define a common naming scheme for all models parameters;
- Define as many common parameters as possible in the base models;
- Implement as many processing steps as possible in the base models;
- For every inheriting model, explain the link between the new "unified" arguments and the arguments from the original implementations/articles.

---

**In a second step**, such a unified framework would allow to easily:
-  Get the embedding (last hidden layer) of any model;
- Indifferently pass a number of samples or a number of seconds for the time-related parameters;
- Improve the skorch EEG classifier and regressor:
    - they could automatically fill some parameters during `fit` according to the training data, such as the number of classes, the temporal length of the input, the sampling frequency (requires `mne.Epochs`) or the number of channels;
    - because they would require a model inheriting from the base Braindecode models, they could simply be initialized with a model name, allowing to programmatically load any Braindecode model.

**Comment by robintibor** [comment]
Interesting, 2 points:
1) Looking a bit how [timm](https://github.com/huggingface/pytorch-image-models) does things may be helpful
2) `fully-connected + softmax at the end (processing step)` Not sure this is actually good though, like often one could absorb softmax into the loss during training as well... also fully connected depends on how you define it, quite  a few models do only have a convolutional layer that is extended to be "fully-connected" depending on the passed arguments.

**Comment by PierreGtch** [comment]
Yes, I'm not a big fan of integrating the softmax in the network either but it is implemented this way right now and it is the expected default for skorch classifiers. Maybe it would be worth changing that.

Indeed, a fully connected layer might not be the best option for all models but it would be nice if at least all the classification layers could share the same name to be able to remove it automatically. 

I will check out timm, thanks!

**Comment by bruAristimunha** [comment]
@apmellot, @sara04, @dengemann, @jpaillard

**Comment by PierreGtch** [comment]
Thanks a lot @brunaafl, @Sara04, @sliwy, @bruAristimunha and @robintibor for your help on this issue!!
